pytential/symbolic/pde/beltrami.py

```diff
@@ -0,0 +1,268 @@
+THE SOFTWARE.
+"""
+
+__doc__ = """
+.. autoclass:: LaplaceBeltramiOperator
+.. autoclass:: YukawaBeltramiOperator
+.. autoclass:: HelmholtzbeltramiOperator
+"""
+
+from functools import partial
+from typing import Any, Dict, Optional
+
+import numpy as np
+
+from pytential import sym
+from sumpy.kernel import Kernel
+
+
+# {{{ beltrami operator
+
+class BeltramiOperator:
+    def __init__(self, kernel: Kernel, *,
+            dim: Optional[int] = None,
+            precond: str = "left",
+            kernel_arguments: Optional[Dict[str, Any]] = None) -> None:
+        if dim is None:
+            dim = kernel.dim - 1
+
+        if precond not in ("left", "right"):
+            raise ValueError(f"unknown preconditioning '{precond}'")
+
+        if kernel_arguments is None:
+            kernel_arguments = {}
+
+        self.dim = dim
+        self.kernel = kernel
+        self.precond = precond
+        self.kernel_arguments = kernel_arguments
+
+    @property
+    def ambient_dim(self):
+        return self.kernel.dim
+
+    @property
+    def dtype(self):
+        return np.dtype(
+                np.complex128 if self.kernel.is_complex_valued
+                else np.float64)
+
+    def get_density_var(self, name: str = "sigma"):
+        return sym.var(name)
+
+    def prepare_solution(self, sigma: sym.var) -> sym.var:
+        S = partial(sym.S, self.kernel,     # noqa: N806
+                qbx_forced_limit=+1, **self.kernel_arguments)
+
+        if self.precond == "left":
+            return S(sigma)
+        else:
+            return S(S(sigma))
+
+    def prepare_rhs(self, b: sym.var) -> sym.var:
+        S = partial(sym.S, self.kernel,     # noqa: N806
+                qbx_forced_limit=+1, **self.kernel_arguments)
+
+        if self.precond == "left":
+            return S(b)
+        else:
+            return b
+
+    def _get_left_operator(self, sigma: sym.var, kappa: sym.var, **kwargs: Any):
+        from sumpy.kernel import LaplaceKernel
+        if isinstance(self.kernel, LaplaceKernel):
+            raise TypeError(
+                    f"{type(self).__name__} does not support the Laplace "
+                    "kernel, use LaplaceBeltramiOperator instead")
+
+        context = self.kernel_arguments.copy()
+        context.update(kwargs)
+        kappa = self.dim * kappa
+
+        knl = self.kernel
+        lknl = LaplaceKernel(knl.dim)
+
+        # {{{ layer potentials
+
+        # laplace
+        S0 = partial(sym.S, lknl, qbx_forced_limit=+1, **kwargs)        # noqa: N806
+        D0 = partial(sym.D, lknl, qbx_forced_limit="avg", **kwargs)     # noqa: N806
+        Dp0 = partial(sym.Dp, lknl, qbx_forced_limit="avg", **kwargs)   # noqa: N806
+
+        # base
+        S = partial(sym.S, knl, qbx_forced_limit=+1, **context)         # noqa: N806
+        Sp = partial(sym.Sp, knl, qbx_forced_limit="avg", **context)    # noqa: N806
+        Spp = partial(sym.Spp, knl, qbx_forced_limit="avg", **context)  # noqa: N806
```

Comment:
Are the `qbx_forced_limit` right here? `Sp` and `D` are probably right, but not sure about the hypersingular ones (here and below). Also for `S`, this only ever evaluates on the surface, so not quite sure if any side is to be prefered?

Reply:
Yep, those look fine to me. `"avg"` gives the principal value/finite part, also for the hypersingular integrals.